roxie/ccd/ccdfile.cpp

```diff
@@ -1270,6 +1621,211 @@ class CRoxieFileCache : implements IRoxieFileCache, implements ICopyFileProgress
+                const char *eol = strchr(cacheInfo, '\n');
+                if (!eol)
+                    break;
+                cacheInfo = eol+1;
+                continue;
+            }
+            cacheInfo++;
+            const char *endName = strchr(cacheInfo, '|');
+            assert(endName);
+            if (!endName)
+                break;
+            StringBuffer fileName(endName-cacheInfo, cacheInfo);
+            Owned<IKeyIndex> keyIndex;
+            bool keyFailed = false;
+            unsigned fileIdx = (unsigned) -1;
+            char *file_mmap = nullptr;
+            int fd = open(fileName, 0);
+            struct stat file_stat;
+            if (fd != -1)
+            {
+                fstat(fd, &file_stat);
+                file_mmap = (char *) mmap((void *)0, file_stat.st_size, PROT_READ, MAP_SHARED, fd, 0);
+                if (file_mmap == MAP_FAILED)
+                {
+                    DBGLOG("Failed to map file %s to pre-warm cache (error %d)", fileName.str(), errno);
+                    file_mmap = nullptr;
+                }
+            }
+            else if (traceLevel)
+            {
+                DBGLOG("Failed to open file %s to pre-warm cache (error %d)", fileName.str(), errno);
+            }
+            Owned<ILazyFileIO> localFile = lookupLocalFile(fileName);
+            if (localFile)
+            {
+                fileIdx = localFile->getFileIdx();
+            }
+            cacheInfo = endName+1;  // Skip the |
+            while (*cacheInfo==' ')
+                cacheInfo++;
+            for (;;)
+            {
+                bool inNodeCache = (*cacheInfo=='*');
+                if (inNodeCache)
+                    cacheInfo++;
+                offset_t startPage = readPage(cacheInfo);
+                offset_t endPage;
+                if (*cacheInfo=='-')
+                {
+                    cacheInfo++;
+                    endPage = readPage(cacheInfo);
+                }
+                else
+                    endPage = startPage;
+                if (traceLevel > 4)
+                    DBGLOG("Touching %s %" I64F "x-%" I64F "x", fileName.str(), startPage, endPage);
+                startPage = startPage << CacheInfoEntry::pageBits;
+                endPage = (endPage+1) << CacheInfoEntry::pageBits;
+                if (inNodeCache && !keyFailed && localFile && !keyIndex)
+                {
+                    keyIndex.setown(createKeyIndex(fileName, localFile->getCrc(), *localFile.get(), fileIdx, false, false));  // MORE - we don't know if it's a TLK, but hopefully it doesn't matter
+                    if (!keyIndex)
+                        keyFailed = true;
+                }
+                if (inNodeCache && keyIndex)
+                {
+                    // Round startPage up to nearest multiple of node size
+                    unsigned nodeSize = keyIndex->getNodeSize();
+                    startPage = ((startPage+nodeSize-1)/nodeSize)*nodeSize;
```

Comment:
am I right in thinking that this is a no-op normally, if the jhtree page is 8k and the prewarm cache page size is also 8k?

Reply:
Normally, but users CAN change the pagesize, and we may want to at some point change the default.